### PR TITLE
Non-repudiation: use binary headers for signature and fingerprint

### PR DESCRIPTION
**Pull request by stefanobaghino-da**
This should reduce the times we encode and decode to and from base64 strings.

changelog_begin
changelog_end

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [x] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: runtime-components/non-repudiation-core/src/main/java/com/daml/nonrepudiation/Headers.java [hunk 1]
```diff
@@ -10,13 +10,13 @@
     private Headers() {
     }
 
-    public static final Metadata.Key<String> SIGNATURE =
-            Metadata.Key.of("signature", Metadata.ASCII_STRING_MARSHALLER);
+    public static final Metadata.Key<byte[]> SIGNATURE =
```
Comment (cocreature) [inline runtime-components/non-repudiation-core/src/main/java/com/daml/nonrepudiation/Headers.java:13]: Is there a good reason to have a separate header name for this at this point given that we don’t care about backwards compat?

File: runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala [hunk 2]
```diff
@@ -5,24 +5,20 @@ package com.daml.nonrepudiation
...
-    def assertFromString(string: String): Fingerprint =
-      string.asInstanceOf[Fingerprint]
-
-    def fromString(string: String): Try[Fingerprint] =
-      Try(assertFromString(string))
-
-    def compute(key: PublicKey): Fingerprint =
-      assertFromString(Base64Fingerprint.compute(key))
-
+  @SuppressWarnings(Array("org.wartremover.warts.ArrayEquals"))
```
Comment (cocreature) [inline runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala:19]: Do we trigger this via some autoderived equals instance? Should we overwrite equality to do something sensible?

**Review by cocreature — APPROVED** [review]
lgtm with one comment on the header names, thanks!

File: runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala:19]: I'll have a look if we can do something about it.

File: runtime-components/non-repudiation-core/src/main/java/com/daml/nonrepudiation/Headers.java (hunk 1 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation-core/src/main/java/com/daml/nonrepudiation/Headers.java:13]: I would have kept it the same, but gRPC actually [throws an exception](https://github.com/grpc/grpc-java/blob/a6df2b2ff403f9aa9097e3fdc6908a53a8f3c4bc/api/src/main/java/io/grpc/Metadata.java#L851-L860) if I don't append `-bin` to the name of binary headers.

File: runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala:19]: Value classes cannot redefine `equals` and I believe `==` is simply synthesized by the compiler. @S11001001 do you have suggestions?

Comment (S11001001) [reply runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala:19]: Do not use a value class for this kind of thing. [They are only good for syntax extension.](https://discuss.daml.com/t/extends-anyval-is-good-for-extension-methods-and-thats-it-scala-ritual/1548)

You can then manually write an `equals` or, if you change to a `case class`, mix in [`SomeArrayEquals`](https://github.com/digital-asset/daml/blob/81dc6b947374886c0ba41fe27fa0546b09659071/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SomeArrayEquals.scala).

If you really want a newtype, use the abstract type with existential implementation pattern we use for all sorts of things around the codebase like `LedgerString`, `NewBoolean`, `com.daml.concurrent.Future`, or `DiscriminatedSubtype`.  No synthetic `equals` implementations are made in those cases; of course this is not much help to you if you _need_ `equals` to work (but this is completely incompatible with using a newtype at all).

Comment (stefanobaghino-da) [reply runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala:19]: I went with the abstract type initially but It accepted a simple `Array[Byte]` in its place. I'll have a second look if I can make it work.

File: runtime-components/non-repudiation-core/src/main/java/com/daml/nonrepudiation/Headers.java (hunk 1 above)
Comment (cocreature) [reply runtime-components/non-repudiation-core/src/main/java/com/daml/nonrepudiation/Headers.java:13]: huh, interesting, makes sense thanks!

File: runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala (hunk 2 above)
Comment (stefanobaghino-da) [reply runtime-components/non-repudiation/src/main/scala/com/daml/nonrepudiation/KeyRepository.scala]: Used it, this time raw `Array[Byte]` are rejected. No idea what I got wrong the first time.

Addressed by https://github.com/digital-asset/daml/pull/8672/commits/a099a8f618b0769b742920281ddc27e345524b2e

**Pull request merged**